### PR TITLE
#195: Add keyfield and keys to schema validation error report

### LINKED ISSUE #195: Add keyField and keys to schema validation error report
New format can be like:
```
"errors": [
        {
          "message": "MUST be string but found null",
          "occurrences": [
            { "count": 1, "keyField": "ListingKey", "keys": ["ABC123"]}
          ]
        }
      ]
```

Refer: [link](https://github.com/RESOStandards/web-api-commander/blob/c528c9061de0cdcd8b1196d22512964d3788af57/src/main/java/org/reso/commander/common/DataDictionaryMetadata.java#L15)

### PR DESCRIPTION
**Pull request by mohit-s96**
closes #195 

also adds `lookupValue` in case the failed enum was an empty string.

### CODE REVIEW

**Comment by darnjo** [comment]
@mohit-s96 - it would probably be better if we could turn off keys for warnings and only use them on errors. The report could get very large.

File: lib/schema/utils.js [hunk 1]
```diff
@@ -389,9 +403,11 @@ const combineErrors = ({ errorCache, warningsCache, payloadErrors, stats, versio
               count: details.occurrences,
               ...(file && { fileName: file }),
               ...(pathName && { pathName }),
-              ...(lookupValue && { lookupValue }),
+              ...(details?.failedEnum != null && { lookupValue }),
```
Comment (darnjo) [inline lib/schema/utils.js:406]: Are you meaning to use loose equality here? And is there a way around it?

File: lib/schema/utils.js [hunk 2]
```diff
@@ -389,9 +403,11 @@ const combineErrors = ({ errorCache, warningsCache, payloadErrors, stats, versio
...
               ...(details?.sourceModel && { sourceModel: details.sourceModel }),
               ...(details?.sourceModelField && { sourceModelField: details.sourceModelField }),
+              ...(details?.keyField && { keyField: details.keyField }),
```
Comment (darnjo) [inline lib/schema/utils.js:409]: Probably better if you don't use fall-through evaluation of `&&` to return the value here.

File: lib/schema/utils.js [hunk 3]
```diff
@@ -499,10 +521,16 @@ const updateCacheAndStats = ({
         : message,
       occurrences: 1,
       sourceModel,
-      sourceModelField
+      sourceModelField,
+      ...(keyField && keyFieldValue && { keyField }),
```
Comment (darnjo) [inline lib/schema/utils.js:525]: Same as previous comment.

File: lib/schema/utils.js [hunk 4]
```diff
@@ -499,10 +521,16 @@ const updateCacheAndStats = ({
...
+      ...(keyFieldValue && { keys: [keyFieldValue] }),
+      ...(failedEnum != null && { failedEnum })
     };
   } else {
     cache[resourceName][failedItemName][message][filePath][failedItemValue].occurrences++;
+    if (cache[resourceName][failedItemName][message][filePath][failedItemValue]?.keys && keyFieldValue) {
```
Comment (darnjo) [inline lib/schema/utils.js:531]: Optional chaining?

File: lib/schema/utils.js [hunk 5]
```diff
@@ -550,6 +578,76 @@ const getMaxLengthMessage = (limitObject, isRCF) => {
   }
 };
 
+const keyifyResourceName = resourceName => {
+  return resourceName.trim() + 'Key';
+};
+
+// prettier-ignore
```
Comment (darnjo) [inline lib/schema/utils.js:585]: Shouldn't do this - what's the error here?

File: lib/schema/validate.js [hunk 6]
```diff
@@ -296,7 +298,7 @@ const generateErrorReport = ({
     }
 
     // eslint-disable-next-line prefer-const
```
Comment (darnjo) [inline lib/schema/validate.js:300]: Any way to make this a const instead? Better not to use these hints - they're there in eslint for a reason. ;)

File: lib/schema/validate.js [hunk 7]
```diff
@@ -360,6 +363,18 @@ const generateErrorReport = ({
       }
     }
 
+    const multiPayload = validationContext.getPayloadType() === 'MULTI';
+    const payload = multiPayload ? json : { value: [json] };
+    const keyField = isWarning ? null : getKeyFieldForResource(sourceModel || resourceName);
```
Comment (darnjo) [inline lib/schema/validate.js:368]: Better to use coalesce.

**Review by darnjo — CHANGES_REQUESTED** [review]

File: lib/schema/utils.js (hunk 1 above)
Comment (mohit-s96) [reply lib/schema/utils.js]: Was meant for checking for either undefined or null but I've made it more explicit now.

**Review by darnjo — APPROVED** [review]

**Pull request merged**